### PR TITLE
feat: add unwrapping feature to OA CLI

### PR DESCRIPTION
**Pull request by zixiang2018**
**Context**
To support unwrapping of a single wrapped document in the OA CLI

**What does this PR do?**
Supports the command below to take in a single wrapped document and unwrap it to a specified folder output. 

Command - `npm run dev -- unwrap <wrapped-document-file> --output-file <output-document-file>`

### CODE REVIEW

**Review by cavacado — COMMENTED** [review]
i've discussed with you separately, so make the relevant changes and i'll review again.

File: src/commands/unwrap.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import { isDir } from "../implementations/utils/disk";
+import { SchemaId } from "@govtechsg/open-attestation";
+
+interface UnwrapCommand {
+  wrappedDocumentsPath: string;
+  outputDir?: string;
+  outputFile?: string;
+  schema?: string;
+  openAttestationV3: boolean;
+  unwrap: boolean;
```
Comment (simboonlong) [inline src/commands/unwrap.ts:14]: i not sure if there is such a thing to unwrap v3

File: src/implementations/unwrap/unwrap.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  Directory,
+  StdOut,
+}
+
+const remoteLoadSchema = async (uri: string): Promise<AnySchemaObject> => {
+  const response = await fetch(uri);
+  console.log({ uri });
+  const r = await response.json();
+  console.log({ uri });
+  return r;
```
Comment (simboonlong) [inline src/implementations/unwrap/unwrap.ts:31]: logs usually can remove after done, try avoid const 1 letter naming kind ya

**Comment by simboonlong** [comment]
might want to update readme + add unit tests too

File: src/commands/unwrap.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      inputPath: args.wrappedDocumentsPath,
+      outputPath,
+      outputPathType,
+    });
+
+    if (rawDocs) {
+      if (rawDocs.length > 1) {
+        signale.success(`The documents have been unwrapped into folder ${outputPath}`);
+      } else {
+        signale.success(`The document have been unwrapped`);
```
Comment (cavacado) [inline src/commands/unwrap.ts:70]: english

File: src/implementations/unwrap/unwrap.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  document: any;
+}): void => {
+  if (outputPathType === Output.File && unwrappedDocumentPath) {
+    writeDocumentToDisk(path.parse(unwrappedDocumentPath).dir, path.parse(unwrappedDocumentPath).base, document);
+  } else if (outputPathType === Output.Directory && unwrappedDocumentPath) {
+    writeDocumentToDisk(unwrappedDocumentPath, path.parse(file).base, document);
+  } else {
+    console.log(JSON.stringify(document, undefined, 2)); // print to console, no file created
+  }
+};
```
Comment (cavacado) [inline src/implementations/unwrap/unwrap.ts:57]: refactor this and the one in implementation/wrap.ts, (move into disk utils) then consume here.

File: src/implementations/unwrap/unwrap.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  outputPathType,
+}: UnwrapArguments): Promise<OpenAttestationDocument[] | undefined> => {
+  // Create output dir
+  if (outputPath) {
+    mkdirp.sync(outputPathType === Output.File ? path.parse(outputPath).dir : outputPath);
+  }
+
+  const unwrappedDocumentsArray = await unwrapIndividualDocuments(inputPath, outputPath, outputPathType);
+
+  return unwrappedDocumentsArray;
```
Comment (cavacado) [inline src/implementations/unwrap/unwrap.ts:77]: return the length of the array here (or at `unwrapIndividualDocs`), then check for the length at the command level to determine the message for the consumer.

File: src/commands/unwrap.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      }
+    }
+  } catch (err) {
+    signale.error(err.message);
+    if (err.validationErrors) {
+      for (const error of transformValidationErrors(err.validationErrors)) {
+        signale.error(error);
+      }
+    }
```
Comment (cavacado) [inline src/commands/unwrap.ts:79]: do we need validation errors? no right?

File: src/implementations/unwrap/unwrap.ts (hunk 5 above)
Comment (zixiang2018) [reply src/implementations/unwrap/unwrap.ts]: I don't know of any use case for when we need the contents in unwrappedDocsArray 😰 . 

Have changed it to return the length instead.

File: src/commands/unwrap.ts (hunk 6 above)
Comment (zixiang2018) [reply src/commands/unwrap.ts]: I removed the validation errors loop for this, will just show the error message instead!

File: src/implementations/unwrap/unwrap.ts (hunk 4 above)
Comment (zixiang2018) [reply src/implementations/unwrap/unwrap.ts:62]: Shifted to [`src/implementations/utils/disk.ts`](https://github.com/Open-Attestation/open-attestation-cli/blob/feat/add-unwrap-command/src/implementations/utils/disk.ts#L43)

File: src/commands/unwrap.ts (hunk 3 above)
Comment (zixiang2018) [reply src/commands/unwrap.ts]: Modified here https://github.com/Open-Attestation/open-attestation-cli/blob/feat/add-unwrap-command/src/commands/unwrap.ts#L69

File: src/implementations/unwrap/unwrap.ts (hunk 4 above)
Comment (cavacado) [reply src/implementations/unwrap/unwrap.ts:62]: i meant the whole `writeOutput` function

File: README.md [hunk 7]
```diff
@@ -151,6 +152,38 @@ open-attestation wrap ./examples/raw-documents/ ./examples/wrapped-documents/ --
...
+### Unwrapping documents
+
+This command processes a document in the input directory. It will unwrap the wrapped document to its raw document form to be displayed on the console.
+
+Example:
+
+```bash
+open-attestation unwrap ./examples/v2/wrapped-documents/example.0.json
+
+✔  success   The document have been unwrapped
```
Comment (cavacado) [inline README.md:164]: eng

File: README.md [hunk 8]
```diff
@@ -151,6 +152,38 @@ open-attestation wrap ./examples/raw-documents/ ./examples/wrapped-documents/ --
...
+```
+
+The command will display the result in the console. If you need to save the file you can use the `--output-file` option.
+
+Example:
+
+```bash
+open-attestation unwrap ./examples/v2/wrapped-documents/example.0.json --output-file ./examples/v2/raw-documents/example.0.json
+
+✔  success   The document have been unwrapped
```
Comment (cavacado) [inline README.md:174]: same

File: src/implementations/unwrap/unwrap.ts (hunk 4 above)
Comment (zixiang2018) [reply src/implementations/unwrap/unwrap.ts:37]: Shifted `writeOutput` function to [`disk.ts` ](https://github.com/Open-Attestation/open-attestation-cli/blob/feat/add-unwrap-command/src/implementations/utils/disk.ts#L48)

File: README.md (hunk 7 above)
Comment (zixiang2018) [reply README.md]: [Updated](https://github.com/Open-Attestation/open-attestation-cli/blob/feat/add-unwrap-command/README.md?plain=1#L164)

File: README.md (hunk 8 above)
Comment (zixiang2018) [reply README.md]: [Updated](https://github.com/Open-Attestation/open-attestation-cli/blob/feat/add-unwrap-command/README.md?plain=1#L174)

**Review by cavacado — APPROVED** [review]
👍🎊🎉

**Pull request merged**